### PR TITLE
Adding the MicroKernel article

### PR DESCRIPTION
**Pull request by weaverryan**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 2.8+ |
| Fixed tickets | #5864 |

Here's the `MicroKernelTrait` article. Finished code behind this can be found here: https://github.com/weaverryan/docs-micro_kernel/tree/finished

This shows the smallest example (single file) and an example that is appropriately large, with the WDT, annotation routes, Twig and a sensible directory structure (which shows off importing external resources, etc). With a few code comments, I think (hope) the code speaks for itself.

Thanks!

### CODE REVIEW

File: cookbook/configuration/micro-kernel-trait.rst [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+When you use the ``MicroKernelTrait``, your kernel needs to have exactly three methods
+that define your bundles, your services and your routes:
+
+**registerBundles()**
+    This is the same ``registerBundles()`` that you see in a normal kernel.
+
+**configureContainer(ContainerBuilder $c, LoaderInterface $loader)**
+    This methods builds and configures the container. In practice, you will use
+    ``loadFromExtension`` to configure different bundles (this is the equivalent)
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]:
```
(this is the equivalent) of what you see in a normal ``config.yml`` file
```

should be like this?

```
(this is the equivalent of what you see in a normal ``config.yml`` file)
```

File: cookbook/configuration/micro-kernel-trait.rst [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    of what you see in a normal ``config.yml`` file. You can also register services
+    directly in PHP or load external configuration files (shown below).
+
+**configureRoutes(RouteCollectionBuilder $routes)**
+    Your job in this method is to add routes to the system. The ``RouteCollectionBuilder``
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]: `to the system` -> `to the application`?

File: cookbook/configuration/micro-kernel-trait.rst [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    is new in Symfony 2.8 and has methods that make adding routes in PHP more fun.
+    You can also load external routing files (shown below).
+
+
+Advanced Example: Twig, Annotations and the Web Debug Toolbar
+-------------------------------------------------------------
+
+The purpose of the ``MicroKernelTrait`` is *not* to have a single-file application.
+Instead, it's goal to give you the power to choose your bundles and structure.
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]: `it's goal` -> `its goal`

File: cookbook/configuration/micro-kernel-trait.rst [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+First, you'll probably want to put your PHP classes in an ``src/`` directory. Configure
+your ``composer.json`` file to load from there:
+
+... code-block:: json
+
+{
+    "require": {
+        "...": "...",
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]: The trailing comma is not valid JSON syntax.

File: cookbook/configuration/micro-kernel-trait.rst [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    use Symfony\Component\Config\Loader\LoaderInterface;
+    use Symfony\Component\DependencyInjection\ContainerBuilder;
+    use Symfony\Component\HttpKernel\Kernel;
+    use Symfony\Component\Routing\RouteCollectionBuilder;
+    use Doctrine\Common\Annotations\AnnotationRegistry;
+
+    // require Composer's autoloader
+    $loader = require __DIR__.'/../vendor/autoload.php';
+    // auto-load annotations
+    AnnotationRegistry::registerLoader(array($loader, 'loadClass'));
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]: Do we really need this line of code? In the tests I did, annotations seemed to work without adding this line.

File: cookbook/configuration/micro-kernel-trait.rst [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+Advanced Example: Twig, Annotations and the Web Debug Toolbar
+-------------------------------------------------------------
+
+The purpose of the ``MicroKernelTrait`` is *not* to have a single-file application.
+Instead, it's goal to give you the power to choose your bundles and structure.
+
+First, you'll probably want to put your PHP classes in an ``src/`` directory. Configure
+your ``composer.json`` file to load from there:
+
+... code-block:: json
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]:
```
... code-block:: json
```

should be:

```
.. code-block:: json
```

File: cookbook/configuration/micro-kernel-trait.rst [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+{
```
Comment (javiereguiluz) [inline cookbook/configuration/micro-kernel-trait.rst]: The code block contents should be indented.

**Comment by javiereguiluz** [comment]
:+1: fantastic article! Thanks.

File: cookbook/configuration/micro-kernel-trait.rst [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+A Single-File Symfony Application
+---------------------------------
+
+Start with a completely empty directory. Get ``symfony/symfony`` as a dependency
+via Composer:
+
+.. code-block:: bash
+
+    composer require symfony/symfony
```
Comment (WouterJ) [inline cookbook/configuration/micro-kernel-trait.rst]: commands always start with `$` and missing version

File: cookbook/configuration/micro-kernel-trait.rst (hunk 5 above)
Comment (weaverryan) [inline cookbook/configuration/micro-kernel-trait.rst]: There is some weird behavior here. Yes, you do need this - I just double-checked by removing that code (and clearing the cache) and I got the error:

>
[Semantical Error] The annotation "@Sensio\Bundle\FrameworkExtraBundle\Configuration\Route"
> in method App\Controller\MicroController::randomAction() does not exist, or could not be auto-loaded.

But, the other day in a different project, I *swear* I removed this, cleared the cache, and *didn't* get that error. For now, I'll include it - but yes, I'd love for this to happen automatically.

**Comment by weaverryan** [comment]
Thanks for the review guys :)

**Pull request merged**